external/fv3fit/fv3fit/reservoir/domain2.py

```diff
@@ -0,0 +1,222 @@
+        if len(data_shape) > len(self._rank_dims_all_features):
+            dim_slices = [..., *dim_slices]
+
+        return dim_slices
+
+    def get_subdomain(self, data, subdomain_index):
+
+        if subdomain_index < 0 or subdomain_index >= self.n_subdomains:
+            raise ValueError(
+                f"Subdomain index {subdomain_index} out of range "
+                "[0, {self.n_subdomains})"
+            )
+
+        _check_feature_dims_consistent(data.shape, self._rank_extent_for_check)
+        dim_slices = self._get_subdomain_slice(subdomain_index)
+        dim_slices = self._add_potential_leading_dim_to_slices(data.shape, dim_slices)
+
+        return data[dim_slices]
+
+    def get_all_subdomains(self, data):
+
+        new_index_dim = -1 * len(self.all_subdomains_shape)
+        subdomains_with_new_dim = []
+        for i in range(self.n_subdomains):
+            subdomain = self.get_subdomain(data, i)
+            subdomain_newaxis = np.expand_dims(subdomain, axis=new_index_dim)
+            subdomains_with_new_dim.append(subdomain_newaxis)
+        return np.concatenate(subdomains_with_new_dim, axis=new_index_dim)
+
+    @property
+    def subdomain_shape(self):
+        shape = list(self.subdomain_xy_extent)
+        if self._z_feature is not None:
+            shape += [self._z_feature]
+        return shape
+
+    @property
+    def flat_subdomain_shape(self):
+        return [np.prod(self.subdomain_shape)]
+
+    @property
+    def all_subdomains_shape(self):
+        return [self.n_subdomains, *self.subdomain_shape]
+
+    def flatten_subdomain_features(self, data):
+        feature_shape = self.subdomain_shape
+        _check_feature_dims_consistent(data.shape, feature_shape)
+        n_feature_dims = len(feature_shape)
+        return data.reshape(list(data.shape[:-n_feature_dims]) + [-1])
+
+    def reshape_flat_subdomain_features(self, data):
+        flat_feature_shape = self.flat_subdomain_shape
+        _check_feature_dims_consistent(data.shape, flat_feature_shape)
+        original_shape = list(data.shape[:-1]) + self.subdomain_shape
+        return data.reshape(original_shape)
+
+    def merge_all_subdomains(self, data):
+        # [leading, nsubdomains, ..., x, y, (z)]
+
+        _check_feature_dims_consistent(data.shape, self.all_subdomains_shape)
+        rank_extent = self._rank_extent_all_features
+        subdomain_axis = -1 * len(self.all_subdomains_shape)
+        new_shape = list(data.shape[:subdomain_axis]) + rank_extent
+        merged = np.empty(new_shape, dtype=data.dtype)
+
+        for i in range(self.n_subdomains):
+            subdomain = np.take(data, i, axis=subdomain_axis)
+            dim_slices = self._get_subdomain_slice(i)
+            dim_slices = self._add_potential_leading_dim_to_slices(
+                subdomain.shape, dim_slices
+            )
+            merged[dim_slices] = subdomain
+
+        return merged
+
+
+class OverlapRankXYDivider(RankXYDivider):
```

Comment:
I opted to separate this out into a subclass to remove the `with_overlap: bool` flags needed to many functions.  However, I'm not attached to this method as it is probably less clear when used within the code, since then we need _two_ rank dividers.  I do have a method that generates a `RankDivider` from the underlying, so we only ever need a single one at the top level.